demos/vlc-dag/src/lib.rs

```diff
@@ -0,0 +1,720 @@
+    pub fn from_file(path: &str) -> Self {
+        let mut config = Configuration {
+            server_addrs: Vec::new(),
+        };
+        let file = std::fs::File::open(path).unwrap();
+        let reader = std::io::BufReader::new(file);
+        for line in reader.lines() {
+            let addr = line.unwrap().parse().unwrap();
+            config.server_addrs.push(addr);
+        }
+        config
+    }
+}
+
+/// Client message type for the accumulator application. Each message contains
+/// a string.
+#[derive(Serialize, Deserialize, Debug)]
+struct ClientMessage {
+    item: String,
+}
+
+/// The current node state, which is a set of strings.
+#[derive(Serialize, Deserialize, Debug, Clone)]
+enum ServerMessage {
+    EventTrigger(EventTrigger),
+    DiffReq(DiffReq),
+    DiffRsp(DiffRsp),
+    ActiveSync(ActiveSync)
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+struct EventTrigger {
+    clock_info: ClockInfo,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+struct DiffReq {
+    to : u128,
+    from_clock: ClockInfo,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+struct DiffRsp {
+    to : u128,
+    diffs: Vec<String>,
+    from: ClockInfo,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+struct ActiveSync {
+    to : u128,
+    diffs: Vec<String>,
+    latest: ClockInfo,
+}
+
+/// A client node for the accumulator application.
+pub struct Client {
+    socket: UdpSocket,
+    config: Configuration,
+}
+
+impl Client {
+    /// Create a new client
+    pub async fn new(config: &Configuration) -> Self {
+        let s = UdpSocket::bind("0.0.0.0:0").await.unwrap();
+        Self {
+            socket: s,
+            config: config.clone(),
+        }
+    }
+
+    /// Disseminate a string to the accumulator network.
+    pub async fn disseminate(&mut self, item: &str, id: usize) {
+        let msg = Message::FromClient(ClientMessage {
+            item: String::from(item),
+        });
+        self.socket
+            .send_to(
+                serde_json::to_string(&msg).unwrap().as_bytes(),
+                self.config.server_addrs[id],
+            )
+            .await
+            .unwrap();
+    }
+
+    /// Terminate a running accumulator server.
+    pub async fn terminate(&mut self, index: usize) {
+        let msg = Message::Terminate;
+        self.socket
+            .send_to(
+                serde_json::to_string(&msg).unwrap().as_bytes(),
+                self.config.server_addrs[index],
+            )
+            .await
+            .unwrap();
+    }
+}
+
+/// State of a server node.
+#[derive(Serialize, Deserialize, Debug, Clone)]
+struct ServerState {
+    clock_info: ClockInfo,
+    id: u128,
+    items: Vec<String>, // BtreeSet or HashSet can't keep event inserted order.
+    clock_to_eventid: HashMap<String, String>, // clock_id to event_id
+    message_id_set: BTreeSet<String>,
+}
+
+impl ServerState {
+    /// Create a new server state.
+    fn new(id: u128) -> Self {
```

Comment:
It's better to implement a `Default()` function

Reply:
Node id is a must parameter to construct a ServerState.